spec/system/planning_applications/constraints_spec.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe "Planning Application Assessment", type: :system do
+  let!(:assessor) { create :user, :assessor, local_authority: @default_local_authority }
+
+  let!(:planning_application) do
+    create :planning_application, :not_started, local_authority: @default_local_authority, constraints: ["Conservation Area"]
+  end
+
+  before do
+    sign_in assessor
+    visit planning_application_path(planning_application)
+    click_button "Constraints"
+  end
+
+  context "Editing constraints" do
+    it "can add and change constraints" do
+      click_link("Update")
+
+      check("Listed building")
+      check("Broads")
+      check("Flood zone 3")
+      uncheck("Conservation Area")
+
+      click_button("Save")
+
+      click_button("Constraints")
+
+      expect(page).to have_content("Listed building")
+      expect(page).to have_content("Broads")
+      expect(page).to have_content("Flood zone 3")
+      expect(page).to have_no_content("Conservation Area")
+
+      click_link("Update")
+
+      expect(page).to have_field("Broads", checked: true)
+      expect(page).to have_field("Flood zone 3", checked: true)
+      expect(page).to have_field("Conservation Area", checked: false)
+      expect(page).to have_field("Safety hazard area", checked: false)
+    end
+
+    it "can remove all constraints" do
+      click_link("Update")
+
+      uncheck("Conservation Area")
+
+      page.all(".govuk-checkboxes__input").each do |checkbox|
+        expect(checkbox).not_to be_checked
+      end
+    end
+
+    it "correctly displays a constraint not in the approved list supplied by the API" do
+      new_constraints = { conservation_area: true, protected_trees: false, bat_habitat: true }.to_json
+      planning_application.update!(constraints: new_constraints)
+      planning_application.reload
+      click_link("Update")
+
+      expect(page).to have_field("bat_habitat", checked: true)
+    end
```

Comment:
How is this working? We're setting constraints directly on the model as a json hash but it's working anyway? Maybe Rails is doing something clever and making this work. Either way, this test doesn't seem quite right!

Reply:
The reason it was passing was that it was rendering the entire hash as a string. You're right - this is now an irrelevant test that should have been removed after the refactor, as we are already checking this scenario by setting the constraints on line 9 of this file. Removing.

Reply:
Removed.